services/storage/handlers/packages.js

```diff
@@ -0,0 +1,509 @@
+        }
+
+        const outgoing = {};
+        for (const dep in value) {
+          const warnings = [];
+          const validated = check.validDependencyName(dep, warnings);
+          if (!validated) {
+            validationError = new Error(warnings.join(', '));
+            break;
+          }
+          const { canonical } = validated;
+
+          // XXX: how do we validate npm-style short deps like `github/bloo`?
+          if (
+            typeof value[dep] !== 'string' ||
+            !semver.validRange(value[dep])
+          ) {
+            validationError = new Error(
+              `invalid semver range in "${key}" for "${dep}": "${value[dep]}"`
+            );
+          }
+
+          outgoing[canonical] = value[dep];
+        }
+
+        formdata[key] = outgoing;
+        break;
+    }
+  });
+
+  let filecount = 0;
+  form.on('part', part => {
+    if (validationError) {
+      part.resume();
+      return;
+    }
+
+    ++filecount;
+    part.on('error', err => {
+      validationError = err;
+    });
+
+    const filename =
+      './' + decodeURIComponent(String(part.filename)).replace(/^\/+/g, '');
+    formdata.files[filename] = context.storage.add(part);
+
+    if (/^\.\/package\/readme(\.(md|markdown))?/i.test(filename)) {
+      const chunks = [];
+      formdata.derivedFiles['./readme.html'] = context.storage.add(
+        part.pipe(
+          new Transform({
+            transform(chunk, enc, ready) {
+              chunks.push(chunk);
+              return ready();
+            },
+            flush(ready) {
+              try {
+                const readme = Buffer.concat(chunks); // TODO: utf16 is important!
+                const md = markdown.toHTML(String(readme));
+                this.push(md);
+              } finally {
+                ready();
+              }
+            }
+          })
+        )
+      );
+    }
+  });
+
+  if (context.request.headers['content-encoding'] === 'deflate') {
+    const pipe = context.request.pipe;
+    context.request.pipe = (...args) => {
+      return pipe.call(context.request, zlib.createInflate()).pipe(...args);
+    };
+  }
+
+  form.parse(context.request);
+  try {
+    await oncomplete;
+    if (validationError) {
+      throw validationError;
+    }
+  } catch (err) {
+    return response.error(err.message, 400);
+  }
+
+  // leaving bundledDeps out of the deps count.
+  if (
+    Object.keys(formdata.dependencies).length +
+      Object.keys(formdata.optionalDependencies).length +
+      Object.keys(formdata.devDependencies).length +
+      Object.keys(formdata.peerDependencies).length >
+    MAX_DEPENDENCIES
+  ) {
+    return response.error(`Exceeded maximum number of dependencies.`, 400);
+  }
+
+  if (filecount > MAX_FILES) {
+    return response.error(
+      `Exceeded maximum number of files in a version.`,
+      400
+    );
+  }
+
+  await Promise.all(
+    Object.keys(formdata.files).map(filename => {
```

Comment:
`Object.values` might reduce some code here

Reply:
Ah, this is a particularly gross bit of code where I'm overwriting the key in the original object with the result of the promise, so I wouldn't be able to use `Object.values` here (since I need the key to thunk out the attribute.)